cpp/downgrades/7ff6a6e53dbcff09d1b9b758b594bc6d17366863/coroutine.ql

```diff
@@ -0,0 +1,18 @@
+ class Function extends @function {
+  string toString() { none() }
+}
+
+class Type extends @type {
+  string toString() { none() }
+}
+
+class Variable extends @variable {
+  string toString() { none() }
+}
+
+from Function func, Type traits, Variable handle, Variable promise
+where
+  coroutine(func, traits) and
+  coroutine_placeholder_variable(handle, 1, func) and
+  coroutine_placeholder_variable(promise, 2, func)
+select func, traits, handle, promise
```

Comment:
I assume the point of this change is that new-style coroutines don't have to have exactly 1 handle and 1 promise.  Can they have zero now?  Can they have more than one?

If they have zero, I can see that this downgrade script is simply going to delete the coroutine, which actually seems fine - we wouldn't have been extracting anything sensible before.

If more than one, we're going to get something of a cartesian product or `coroutine` rows here.  I'm guessing we only expect small numbers, but I want to draw attention to this possible behaviour.

Reply:
>I assume the point of this change is that new-style coroutines don't have to have exactly 1 handle and 1 promise. Can they have zero now? Can they have more than one?

They will have exactly one of each. The problem with the previous scheme is that it's not easy to add additional variables, which is what we need here. Adding an extra column to `coroutine` makes it difficult to write an upgrade script, as I don't have anything I can sensibly put in the additional column.

Reply:
Makes sense, and nullifies my concern above - thanks.